gradle/scripts/lib/common-info.gradle

```diff
@@ -73,20 +73,32 @@ allprojects {
                 return null
             }
 
-            // Use the overridden one if available.
-            def overriddenAutomaticModuleName = findOverridden('automaticModuleNameOverrides', project)
-            if (overriddenAutomaticModuleName != null) {
-                return overriddenAutomaticModuleName
-            }
+            return project.provider {
+                // per-project override
+                if (project.ext.has('automaticModuleNameOverride')) {
+                    def override = project.ext.get('automaticModuleNameOverride')
+                    if (!(override instanceof String)) {
+                        throw new IllegalStateException("project.ext.automaticModuleNameOverride must be a String: ${override}")
+                    }
+                    return override
+                }
+
+                // Use the overridden one if available.
 
-            // Generate from the groupId and artifactId otherwise.
-            def groupIdComponents = String.valueOf(rootProject.group).split("\\.").toList()
-            def artifactIdComponents =
-                    String.valueOf(project.ext.artifactId).replace('-', '.').split("\\.").toList()
-            if (groupIdComponents.last() == artifactIdComponents.first()) {
-                return String.join('.', groupIdComponents + artifactIdComponents.drop(1))
-            } else {
-                return String.join('.', groupIdComponents + artifactIdComponents)
+                def overriddenAutomaticModuleName = findOverridden('automaticModuleNameOverrides', project)
+                if (overriddenAutomaticModuleName != null) {
+                    return overriddenAutomaticModuleName
+                }
+
+                // Generate from the groupId and artifactId otherwise.
+                def groupIdComponents = String.valueOf(rootProject.group).split("\\.").toList()
+                def artifactIdComponents =
+                        String.valueOf(project.ext.artifactId).replace('-', '.').split("\\.").toList()
+                if (groupIdComponents.last() == artifactIdComponents.first()) {
+                    return String.join('.', groupIdComponents + artifactIdComponents.drop(1))
+                } else {
+                    return String.join('.', groupIdComponents + artifactIdComponents)
```

Comment:
> When checking the specification, it seems like a identifier needs to start with a JavaLetter.
ref: [docs.oracle.com/javase/specs/jls/se16/html/jls-7.html#jls-7.4](https://docs.oracle.com/javase/specs/jls/se16/html/jls-7.html#jls-7.4)


What do you think of normalizing Automatic-Module-Name by using `-` instead of `.` if the first character of an  artifactIdComponent starts with a number?
It seems that the generated names for Scala modules violate the specification.
```

```

Reply:
My initial thought was that we can't rewrite to support all of the rules defined by the jls. Having said this, assuming we won't be adding modules frequently I guess a simple rule is fine. Updated